File: cryptoheir-rs/src/commands/sign.rs
//! Sign command - signs prepared transactions offline (no network required)

use crate::{crypto, qr, tui, types::*, Result};
use tracing::{info, warn};

pub async fn execute(
    input: String,
    output: String,
    qr_input: Option<String>,
    generate_qr: bool,
    skip_review: bool,
) -> Result<()> {
    info!("Loading transaction parameters...");

    // Load tx params from file or QR code
    let tx_params_json = if let Some(qr_file) = qr_input {
        info!("Scanning QR code from {}...", qr_file);
        qr::scan_qr(&qr_file)?
    } else {
        std::fs::read_to_string(&input)?
    };

    let tx_params: TxParams = serde_json::from_str(&tx_params_json)?;

    // Verify transaction is prepared but not signed
    if !tx_params.metadata.prepared {
        return Err(eyre::eyre!("Transaction has not been prepared"));
    }
    if tx_params.metadata.signed {
        warn!("Transaction appears to already be signed");
    }

    // Validate gas parameters match transaction type
    match tx_params.transaction.tx_type {
        2 => {
            // EIP-1559 transaction must have max fees, not gas price
            if tx_params.transaction.max_fee_per_gas.is_none() {
                return Err(eyre::eyre!(
                    "EIP-1559 transaction (type 2) requires maxFeePerGas but it is missing"
                ));
            }
            if tx_params.transaction.max_priority_fee_per_gas.is_none() {
                return Err(eyre::eyre!(
                    "EIP-1559 transaction (type 2) requires maxPriorityFeePerGas but it is missing"
                ));
            }
            if tx_params.transaction.gas_price.is_some() {
                warn!("EIP-1559 transaction (type 2) should not have gasPrice field - it will be ignored");
            }
        }
        0 => {
            // Legacy transaction must have gas price, not EIP-1559 fees
            if tx_params.transaction.gas_price.is_none() {
                return Err(eyre::eyre!(
                    "Legacy transaction (type 0) requires gasPrice but it is missing"
                ));
            }
            if tx_params.transaction.max_fee_per_gas.is_some() || tx_params.transaction.max_priority_fee_per_gas.is_some() {
                warn!("Legacy transaction (type 0) should not have EIP-1559 fee fields - they will be ignored");
            }
        }
        _ => {
            return Err(eyre::eyre!(
                "Unsupported transaction type: {}. Only type 0 (legacy) and type 2 (EIP-1559) are supported",
                tx_params.transaction.tx_type
            ));
        }
    }

    info!(
        "Transaction type {} validated with appropriate gas parameters",
        tx_params.transaction.tx_type
    );

    // Display transaction for review (interactive TUI or simple display)
    if !skip_review {
        info!("Launching transaction review UI...");
        let approved = tui::review_transaction(&tx_params)?;
        if !approved {
            println!("\n✗ Transaction signing cancelled by user");
            return Ok(());
        }
    } else {
        warn!("Skipping interactive review (--skip-review flag)");
        display_transaction_summary(&tx_params);
    }

    // Load private key from environment
    let config = Config::load()?;
    let private_key = config
        .private_key
        .ok_or_else(|| eyre::eyre!("PRIVATE_KEY not set in environment"))?;

    info!("Signing transaction...");

    // Sign the transaction
    let signed_tx = crypto::sign_transaction(&tx_params, &private_key).await?;

    // Save to file
    let json = serde_json::to_string_pretty(&signed_tx)?;
    std::fs::write(&output, &json)?;

    println!("\n✓ Transaction signed successfully!");
    println!("  Output: {}", output);
    println!("  TX Hash: {}", signed_tx.tx_hash);
    println!("  From: {}", signed_tx.from);

    if let Some(addr) = signed_tx.predicted_contract_address {
        println!("  Contract Address: {}", addr);
    }

    // Generate QR code if requested
    if generate_qr {
        info!("Generating QR code...");
        qr::display_qr(&json)?;
    }

    println!("\nNext step: Transfer {} to online machine and run:", output);
    println!("  cryptoheir-rs broadcast -i {}", output);

    Ok(())
}

fn display_transaction_summary(tx_params: &TxParams) {
    println!("\n========== Transaction Review ==========");
    println!("Network: {} (chain ID: {})",
        tx_params.metadata.network.name,
        tx_params.metadata.network.chain_id
    );
    println!("Mode: {:?}", tx_params.mode);
    if let Some(fn_name) = &tx_params.function_name {
        println!("Function: {}", fn_name);
    }
    println!("From: {}", tx_params.transaction.from);
    if let Some(to) = tx_params.transaction.to {
        println!("To: {}", to);
    }
    println!("Nonce: {}", tx_params.transaction.nonce);
    println!("Gas Limit: {}", tx_params.transaction.gas_limit);
    println!("Estimated Cost: {} ETH", tx_params.metadata.estimated_cost);
    println!("========================================");
}

File: cryptoheir-go/internal/commands/sign.go
package commands

import (
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/cryptoheirdotio/cryptoheir/cryptoheir-go/internal/crypto"
	"github.com/cryptoheirdotio/cryptoheir/cryptoheir-go/internal/tui"
	"github.com/cryptoheirdotio/cryptoheir/cryptoheir-go/internal/types"
	"github.com/spf13/cobra"
)

// SignCmd represents the sign command
var SignCmd = &cobra.Command{
	Use:   "sign",
	Short: "Sign a prepared transaction offline",
	Long: `Sign a prepared transaction with your private key.

This command is designed to be run on an air-gapped (offline) machine.
It will display the transaction details in an interactive TUI for review before signing.`,
	RunE: runSign,
}

var (
	signInputFlag      string
	signOutputFlag     string
	signSkipReviewFlag bool
)

func init() {
	SignCmd.Flags().StringVarP(&signInputFlag, "input", "i", "tx-params.json", "Input transaction parameters file")
	SignCmd.Flags().StringVarP(&signOutputFlag, "output", "o", "signed-tx.json", "Output signed transaction file")
	SignCmd.Flags().BoolVar(&signSkipReviewFlag, "skip-review", false, "Skip interactive TUI review (not recommended)")
}

func runSign(cmd *cobra.Command, args []string) error {
	// Load transaction parameters
	txParamsData, err := os.ReadFile(signInputFlag)
	if err != nil {
		return fmt.Errorf("failed to read input file: %w", err)
	}

	var txParams types.TxParams
	if err := json.Unmarshal(txParamsData, &txParams); err != nil {
		return fmt.Errorf("failed to parse transaction parameters: %w", err)
	}

	log.Info("Transaction parameters loaded")
	log.Info("  Network",
		"network", txParams.Metadata.Network.Name,
		"chain_id", txParams.Transaction.ChainID)
	log.Info("  Mode", "mode", txParams.Mode)
	if txParams.FunctionName != "" {
		log.Info("  Function", "function", txParams.FunctionName)
	}

	// Validate transaction parameters
	if err := validateTxParams(&txParams); err != nil {
		return fmt.Errorf("invalid transaction parameters: %w", err)
	}

	// Interactive TUI review (unless skipped)
	if !signSkipReviewFlag {
		log.Info("Launching interactive transaction review...")
		log.Info("(Use --skip-review flag to bypass this step)")

		approved, err := tui.ReviewTransaction(&txParams)
		if err != nil {
			return fmt.Errorf("TUI error: %w", err)
		}

		if !approved {
			log.Info("Transaction signing cancelled by user")
			return nil
		}

		log.Info("✓ Transaction approved by user")
	} else {
		log.Warn("⚠ WARNING: Skipping transaction review (use with caution!)")
	}

	// Load private key from environment
	config, err := types.LoadConfig()
	if err != nil {
		return fmt.Errorf("failed to load config: %w", err)
	}

	if config.PrivateKey == "" {
		return fmt.Errorf("PRIVATE_KEY not set in environment")
	}

	// Sign transaction
	log.Info("Signing transaction...")
	signedTx, err := crypto.SignTransaction(&txParams, config.PrivateKey)
	if err != nil {
		return fmt.Errorf("failed to sign transaction: %w", err)
	}

	// Update metadata
	signedTx.Metadata.SignedAt = time.Now().UTC().Format(time.RFC3339)

	log.Info("✓ Transaction signed successfully")
	log.Info("  TX Hash", "hash", signedTx.TxHash.Hex())
	if signedTx.PredictedContractAddress != nil {
		log.Info("  Predicted Contract Address", "address", signedTx.PredictedContractAddress.Hex())
	}

	// Save signed transaction
	signedData, err := json.MarshalIndent(signedTx, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to serialize signed transaction: %w", err)
	}

	if err := os.WriteFile(signOutputFlag, signedData, 0644); err != nil {
		return fmt.Errorf("failed to write output file: %w", err)
	}

	log.Info("✓ Signed transaction saved", "file", signOutputFlag)
	log.Info("  Next",
		"instruction", fmt.Sprintf("Transfer to online machine and run 'cryptoheir broadcast -i %s --network %s'",
			signOutputFlag, txParams.Metadata.Network.Name))

	return nil
}

// validateTxParams validates transaction parameters before signing
func validateTxParams(txParams *types.TxParams) error {
	// Check gas parameters match transaction type
	if txParams.Transaction.TxType == 2 {
		// EIP-1559
		if txParams.Transaction.MaxFeePerGas == nil || txParams.Transaction.MaxPriorityFeePerGas == nil {
			return fmt.Errorf("EIP-1559 transaction requires max_fee_per_gas and max_priority_fee_per_gas")
		}
	} else if txParams.Transaction.TxType == 0 {
		// Legacy
		if txParams.Transaction.GasPrice == nil {
			return fmt.Errorf("legacy transaction requires gas_price")
		}
	} else {
		return fmt.Errorf("unsupported transaction type: %d", txParams.Transaction.TxType)
	}

	// Validate addresses
	if txParams.Transaction.From.Hex() == "0x0000000000000000000000000000000000000000" {
		return fmt.Errorf("invalid from address")
	}

	// Validate gas limit
	if txParams.Transaction.GasLimit == nil || txParams.Transaction.GasLimit.ToBigInt().Sign() <= 0 {
		return fmt.Errorf("invalid gas limit")
	}

	return nil
}
